language/documentation/hackathon-tutorial/README.md

```diff
@@ -0,0 +1,599 @@
+```
+
+
+#### Method `deposit`
+
+
+The signature of the method `deposit` is given below:
+
+```
+fun deposit<CoinType>(addr: address, check: Coin<CoinType>) acquires Balance
+```
+
+The method deposits the `check` into `addr`. The specification is defined below:
+
+```
+    spec deposit {
+        let balance = global<Balance<CoinType>>(addr).coin.value;
+        let check_value = check.value;
+
+        aborts_if !exists<Balance<CoinType>>(addr);
+        aborts_if balance + check_value > MAX_U64;
+
+        let post balance_post = global<Balance<CoinType>>(addr).coin.value;
+        ensures balance_post == balance + check_value;
+    }
+```
+
+`balance` represents the number of tokens in `addr` before execution and `check_value` represents the number of tokens to be deposited. The method would abort if 1) `addr` does not have the resource `Balance<CoinType>` or 2) the sum of `balance` and `check_value` is greater than the maxium value of the type `u64`. The functional property checks that the balance is correctly updated after the execution.
+
+
+#### Method `transfer`
+
+The signature of the method `transfer` is given below:
+
+```
+public fun transfer<CoinType: drop>(from: &signer, to: address, amount: u64, _witness: CoinType) acquires Balance
+```
+
+The method transfers the `amount` of coin from the account of `from` to the address `to`.
+
+```
+spec transfer {
+        let addr_from = Signer::address_of(from);
+
+        let balance_from = global<Balance<CoinType>>(addr_from).coin.value;
+        let balance_to = global<Balance<CoinType>>(to).coin.value;
+
+        let post balance_from_post = global<Balance<CoinType>>(addr_from).coin.value;
+        let post balance_to_post = global<Balance<CoinType>>(to).coin.value;
+
+        ensures addr_from != to ==> balance_from_post == balance_from - amount;
+        ensures addr_from != to ==> balance_to_post == balance_to + amount;
+        ensures addr_from == to ==> balance_from_post == balance_from;
+    }
+```
+
+The function `Signer::address_of` is called to obtain the address of `from`. Then the balances of `addr_from` and `to` before and after the execution are obtained. In the three `ensures` clauses, `p ==> q` is used to represented the logical implication between p and q. If the source and the target addresses are the same, the balance remains the same. Otherwise, `amount` is deducted from `addr_from` and added to `to`. The aborts conditions are left as an exercise.
+
+
+### Exercises
+- Implement the aborts_if conditions for the `transfer` method.
+- Implement the specification for the `mint` and `publish_balance` method.
+
+The solution to this exercise can be found in `step_7_sol`.
+
+### Step 8: Formally verify my BasicCoin module using Move Prover
+
+
+We can use the command `mpm -p <path/to/BasicCoin> prove` to prove properties for the BasicCoin module. More prover options can be found [here](https://github.com/diem/diem/blob/main/language/move-prover/doc/user/prover-guide.md).
```

Comment:
Why not just `mpm prove` like elsewhere here? Adding the `-p` option may create confusion as it looks like it would be needed.